### PR TITLE
feat(ui-library): radio group - use slots instead of options array

### LINKED ISSUE #507: RadioGroup - use slots to have Radios in RadioGroup & add API to change/reset checked values of radios
#### Description
After deciding to use slots in #354 , we need to adjust the radio group component, so it also uses slots instead of an array.

Similar to design, where the radio component already lives on the same page and is only a part of the radio group, the radio would then just be a part of the Radio Group component in code.

----
#### Pre-Requisites
- [ ] #1085 

----
#### Acceptance Criteria
- [x] The Radio Group contains Radio components inside slots.
- [ ] The Radio Group has an API to read, change or reset the checked values of the radios inside of the group
- [ ] The `blr-radio` component notifies any parties subscribed to the `blrSelectedValueChange` event for changes made by the `blr-radio-group` component
- [ ] Props documentation has been updated in Storybook and in [Props Excel](https://ts.accenture.com/:x:/r/sites/b01ler-core/Shared%20Documents/Documentation_Props.xlsx?d=wb2cc9fb85c3940499652a4e6bbe09ef1&csf=1&web=1&e=dE1hPa) (on the All Props page and on the Components page) > ask @angelicahoyss or @thrbnhrtmnn for support here
- [x] 'options' prop has been removed from the component code
- [ ] The library builds successfully, and there are no reported issues related to the removed property name

----
#### Background information
- During implementation we found out that #765 might need to be fixed as well when doing this, otherwise the changes of this issue can not be tested. Please check #765 again after this ticket is merged.
- [PR from David](https://github.com/deven-org/boiler/pull/899) that crossed the Radio Group when working on #645 , before we decided to leave out the Radio Group here

### PR DESCRIPTION
**Pull request by bar-tay**
closes #507

### CODE REVIEW

**Review by faselbaum — COMMENTED** [review]
Hey @bar-tay 👋 

I looked into the ACs as well as the changes and i found that we're not really using slots the way they are intended here.

We are still not using the `blr-radio` component inside the radio group but instead render custom input elements. We want to get rid of the custom radio rendering inside `blr-radio-group`. 

There are two main problems that arise from this that we should IMO tackle here:

### Problem 1 - How to render `blr-radio` components?

I see two ways we can handle this. Both of which require some research and experimentation but in both cases we should expect users of `blr-radio-group` to pass `blr-radio` elements as children. This needs to be documented.

#### A) Create near perfect clones of the slotted elements
This would require cloning and observing any changes to the source elements via MutationObservers and probably modify source elements, defining trackable get/set accessors for all properties (if that's even possible) to proxy any changes to our clones.

#### B) Use the slotted elements directly
We could just display the slotted elements.

### Problem 2 - Radio state manipulation and change interception
The `input type="radio"` elements inside our `blr-radio` component will lose some of their internal behaviour as they are isolated from one another due to shadow-dom encapsulation.

One of those behaviours we lose, but which is needed for a `radio group`, is changing the state of other radio's of the same group to "off" (defined by the `name=""` attribute) when the checked state changes to "on".

In short: Click one radio in the group. Other radios will automatically turn off.

We need to re-implement this behaviour with JS inside of `blr-radio-group`.

For this to work we need to ensure that `blr-radio` correctly updates the value of the `checked` property when the user clicks on the element. This is currently not the case. 

### Further considerations (Not necessarily part of this PR)
- When considering Problem 1 - Solution B: How do we know when a JS program changes the `checked` state of one of the `blr-radio` elements? We don't fire any events in this case. In this case our `blr-radio-group` would get out of sync and possibly have two or more radios selected simultaneously

**Comment by angsherpa456** [comment]
Hi @bar-tay JFYI when you are working further on this. The second commit https://github.com/deven-org/boiler/commit/6a5c68ab0b90ff7bed906c102666030ccd9522cf. achieves the desired behaviour based on the click event (The wrapper of the slot in radio-group component listens to the event bubbled by the slotted element[radio-button]). However, the following part still needs to be solved.

"When considering Problem 1 - Solution B: How do we know when a JS program changes the checked state of one of the blr-radio elements? We don't fire any events in this case. In this case our blr-radio-group would get out of sync and possibly have two or more radios selected simultaneously"

File: packages/ui-library/src/components/radio/index.ts [hunk 1]
```diff
@@ -96,14 +122,15 @@ export class BlrRadio extends LitElementCustom {
...
+      const id = calculateOptionId(this.label);
       return html`
         <div class="blr-radio ${classes}">
           <input
-            id=${this.optionId || nothing}
+            id=${id || nothing}
             class="${classes} input-control"
             type="radio"
             name=${this.name}
+            .value=${this.value}
```
Comment (faselbaum) [inline packages/ui-library/src/components/radio/index.ts:133]: in addition to the value we should probably also set `checked` property of the native input element.
my guess is that this is what prevented correctly updating the element state when the radio's custom `checked` property changes.

`.checked=${this.checked}`

line 137 only sets the html-attribute which is only used during element initialisation and will not update the input state if changed.

File: packages/ui-library/src/components/radio-group/index.ts [hunk 2]
```diff
@@ -6,38 +6,122 @@ import { SizelessIconType } from '@boiler/icons';
...
+
+      if (label === (event?.target as HTMLInputElement).id) {
+        radioInputElement.checked = true;
+      }
+    });
+
+    event.preventDefault();
+  };
+
+  firstUpdated() {
```
[the remark this answers is not in the document]
Comment (angsherpa456) [reply packages/ui-library/src/components/radio-group/index.ts:81]: Thanks for the hint, adjusted now with the slotchange, setTimeout is now removed.

File: packages/ui-library/src/components/radio-group/index.ts [hunk 3]
```diff
@@ -6,38 +6,122 @@ import { SizelessIconType } from '@boiler/icons';
...
+    const slot = this.renderRoot?.querySelector('slot');
+    this._radioElements = slot?.assignedElements({ flatten: false });
+
+    setTimeout(() => {
+      {
+        this._radioElements?.forEach((item, index, elements) => {
+          if (item.localName !== RADIO_TAG_NAME) {
+            throw new Error('child component of blr-radio-group must be blr-radio');
```
[the remark this answers is not in the document]
Comment (angsherpa456) [reply packages/ui-library/src/components/radio-group/index.ts]: Optinal A seems to be robust and implemented it now. For now its optimal to stick with our blr-radio button only, later on if there is necessity then we can think about it because allowing the input field other than blr-radio adds the extra complexity.

File: packages/ui-library/src/components/radio-group/index.ts [hunk 4]
```diff
@@ -6,38 +6,122 @@ import { SizelessIconType } from '@boiler/icons';
...
+          }
+
+          const radioElement = <HTMLInputElement>item.shadowRoot?.querySelector('input');
```
[the remark this answers is not in the document]
Comment (angsherpa456) [reply packages/ui-library/src/components/radio-group/index.ts]: I agree, however, in this case completely refraining is not possible because of the styling in case of error or size-variant. Nonetheless, with the second adjustment we now need to access the shadow-dom minimum i.e. in only one place during the component initialization which I would say it pretty ok.

File: packages/ui-library/src/components/radio-group/index.ts [hunk 5]
```diff
@@ -6,38 +6,122 @@ import { SizelessIconType } from '@boiler/icons';
...
+                createBlrRadioGroupValueChangeEvent({
+                  originalEvent: { ...new Event('change', { bubbles: true, cancelable: true }), target: radioElement },
+                  selectedValue: radioElement.value,
+                })
+              );
+            },
+            get: function () {
+              return this.hasAttribute('checked');
+            },
+          });
```
[the remark this answers is not in the document]
Comment (angsherpa456) [reply packages/ui-library/src/components/radio-group/index.ts]: Logic now moved to the radio component rather than in the parent component. Changes made to the existing checked property from the developer console is not picked up by any lifecycle or lit methods so explicitly looking out for checked property is necessary.

File: packages/ui-library/src/components/radio/index.ts (hunk 1 above)
Comment (angsherpa456) [reply packages/ui-library/src/components/radio/index.ts:150]: Same reason as above, implementing with .checked=${this.checked} still does not trigger any method when checked is changed from developer console.

**Review by thrbnhrtmnn — COMMENTED** [review]
I have found one small thing in Storybook which seems not to appear on current develop.

When hasHint = true and I also switch hasError to true, it seems like the margin between radios and caption group increases. Further inspections make me think, that the reason is that the radio slots increase in height when hasError = true.

**Review by thrbnhrtmnn — APPROVED** [review]
Hey @angsherpa456 , I checked the component and all points except the one we discussed will be a separate task (first in the list) have been solved.

**Comment by angsherpa456** [comment]
Hi @thrbnhrtmnn here is the issue as separate task for the first in the list https://github.com/deven-org/boiler/issues/1138

**Review by angsherpa456 — APPROVED** [review]
